db/database.go

```diff
@@ -1013,60 +1009,31 @@ func (c *DatabaseCollection) processForEachDocIDResults(ctx context.Context, cal
 // Returns the IDs of all users and roles, including deleted Roles
 func (db *DatabaseContext) AllPrincipalIDs(ctx context.Context) (users, roles []string, err error) {
 
-	if !db.IsServerless() || db.Options.UseViews {
+	if db.UseLegacySyncDocsIndex() || db.Options.UseViews {
 		return db.getAllPrincipalIDsSyncDocs(ctx)
 	}
 
-	// If running in Serverless mode, we can leverage `users` and `roles` index
-	// to fetch users and roles
-	usersCh := db.getUserNamesInBackground(ctx)
-	rolesCh := db.getRoleIDsInBackground(ctx)
-
-	userData := <-usersCh
-	if userData.err != nil {
-		return nil, nil, userData.err
-	}
-	users = userData.value
-
-	rolesData := <-rolesCh
-	if rolesData.err != nil {
-		return nil, nil, rolesData.err
-	}
-	roles = rolesData.value
-
-	return users, roles, err
-}
-
-// used to send users/roles data from background fetch
-type data struct {
-	value []string
-	err   error
-}
-
-func (db *DatabaseContext) getUserNamesInBackground(ctx context.Context) <-chan data {
-	ch := make(chan data, 1)
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	var getUsersErr error
 	go func() {
-		defer close(ch)
-		users, err := db.GetUserNames(ctx)
-		ch <- data{
-			value: users,
-			err:   err,
-		}
+		defer wg.Done()
+		users, getUsersErr = db.GetUserNames(ctx)
 	}()
-	return ch
-}
-
-func (db *DatabaseContext) getRoleIDsInBackground(ctx context.Context) <-chan data {
-	ch := make(chan data, 1)
+	var getRolesErr error
 	go func() {
-		defer close(ch)
-		roles, err := db.getRoleIDsUsingIndex(ctx, true)
-		ch <- data{
-			value: roles,
-			err:   err,
-		}
+		defer wg.Done()
+		includeDeleted := true
+		roles, getRolesErr = db.getRoleIDsUsingIndex(ctx, includeDeleted)
```

Comment:
NIT: any reason why we don't just pass true in instead of using includeDeleted?

Reply:
I have a preference for defining constants so this code is easier to read, since `true` doesn't tell use what the argument would be. Probably doesn't matter much, but this is reasoning.

This is an ancient recommendation now, but this idea comes from `Code Complete` book from Steve McConnell.